# Modifications to DroidCalib
#   upstream: https://github.com/boschresearch/DroidCalib
#   upstream licence: AGPL-3.0
# Modified by the HuRo authors, 2026. This diff is derived from the upstream above and
# is governed by that licence -- NOT by the Apache-2.0 in ../LICENSE, which covers only
# HuRo's own code. See ../THIRD_PARTY_NOTICES.md.

diff --git a/droid_slam/droid.py b/droid_slam/droid.py
index 911cc2f..c7a42ec 100644
--- a/droid_slam/droid.py
+++ b/droid_slam/droid.py
@@ -47,13 +47,13 @@ class Droid:
             self.visualizer.start()
 
         # post processor - fill in poses for non-keyframes
-        self.traj_filler = PoseTrajectoryFiller(self.net, self.video)
+        #self.traj_filler = PoseTrajectoryFiller(self.net, self.video)
 
 
     def load_weights(self, weights):
         """ load trained model weights """
 
-        print(weights)
+        #print(weights)
         self.net = DroidNet()
         state_dict = OrderedDict([
             (k.replace("module.", ""), v) for (k, v) in torch.load(weights).items()])
@@ -85,15 +85,15 @@ class Droid:
         del self.frontend
 
         torch.cuda.empty_cache()
-        print("#" * 32)
+        #print("#" * 32)
         self.backend(7)
 
         torch.cuda.empty_cache()
-        print("#" * 32)
+        #print("#" * 32)
         self.backend(12)
 
-        camera_trajectory = self.traj_filler(stream)
+        #camera_trajectory = self.traj_filler(stream)
         intr = self.video.intrinsics[0, :].clone()
         intr[:4] *= 8.0
-        return camera_trajectory.inv().data.cpu().numpy(), intr.data.cpu().numpy()
+        return None, intr.data.cpu().numpy()
 
diff --git a/droid_slam/factor_graph.py b/droid_slam/factor_graph.py
index eb605d6..e02f088 100644
--- a/droid_slam/factor_graph.py
+++ b/droid_slam/factor_graph.py
@@ -257,7 +257,7 @@ class FactorGraph:
         corr_op = AltCorrBlock(self.video.fmaps.view(1, num*rig, ch, ht, wd))
 
         for step in range(steps):
-            print("Global BA Iteration #{}".format(step+1))
+            #print("Global BA Iteration #{}".format(step+1))
             with torch.cuda.amp.autocast(enabled=False):
                 coords1, mask = self.video.reproject(self.ii, self.jj)
                 motn = torch.cat([coords1 - self.coords0, self.target - coords1], dim=-1)
diff --git a/droid_slam/motion_filter.py b/droid_slam/motion_filter.py
index 5cb18d9..c0171c7 100644
--- a/droid_slam/motion_filter.py
+++ b/droid_slam/motion_filter.py
@@ -42,7 +42,7 @@ class MotionFilter:
 
     @torch.cuda.amp.autocast(enabled=True)
     @torch.no_grad()
-    def track(self, tstamp, image, depth=None, intrinsics=None):
+    def track(self, tstamp, image, depth=None, intrinsics=None, custom=False):
         """ main update operation - run on every frame in video """
 
         Id = lietorch.SE3.Identity(1,).data.squeeze()
@@ -57,9 +57,10 @@ class MotionFilter:
         gmap = self.__feature_encoder(inputs)
 
         ### always add first frame to the depth video ###
-        if self.video.counter.value == 0:
+        if (not custom and self.video.counter.value == 0) or (custom and tstamp==0):
             net, inp = self.__context_encoder(inputs[:,[0]])
             self.net, self.inp, self.fmap = net, inp, gmap
+            if custom: return True
             self.video.append(tstamp, image[0], Id, 1.0, depth, intrinsics / 8.0, gmap, net[0,0], inp[0,0])
 
         ### only add new frame if there is enough motion ###
@@ -69,91 +70,15 @@ class MotionFilter:
             corr = CorrBlock(self.fmap[None,[0]], gmap[None,[0]])(coords0)
 
             # approximate flow magnitude using 1 update iteration
-            _, delta, weight = self.update(self.net[None], self.inp[None], corr)
+            _, delta, _ = self.update(self.net[None], self.inp[None], corr)
 
-            # check motion magnitue / add new frame to video
+            # check motion magnitude / add new frame to video
             if delta.norm(dim=-1).mean().item() > self.thresh:
                 self.count = 0
                 net, inp = self.__context_encoder(inputs[:,[0]])
                 self.net, self.inp, self.fmap = net, inp, gmap
+                if custom: return True
                 self.video.append(tstamp, image[0], None, None, depth, intrinsics / 8.0, gmap, net[0], inp[0])
-
             else:
                 self.count += 1
-
-
-
-
-# class MotionFilter:
-#     """ This class is used to filter incoming frames and extract features """
-
-#     def __init__(self, net, video, thresh=2.5, device="cuda:0"):
-        
-#         # split net modules
-#         self.cnet = net.cnet
-#         self.fnet = net.fnet
-#         self.update = net.update
-
-#         self.video = video
-#         self.thresh = thresh
-#         self.device = device
-
-#         self.count = 0
-
-#         # mean, std for image normalization
-#         self.MEAN = torch.as_tensor([0.485, 0.456, 0.406], device=self.device)[:, None, None]
-#         self.STDV = torch.as_tensor([0.229, 0.224, 0.225], device=self.device)[:, None, None]
-        
-#     @torch.cuda.amp.autocast(enabled=True)
-#     def __context_encoder(self, image):
-#         """ context features """
-#         x = self.cnet(image)
-#         net, inp = self.cnet(image).split([128,128], dim=2)
-#         return net.tanh().squeeze(0), inp.relu().squeeze(0)
-
-#     @torch.cuda.amp.autocast(enabled=True)
-#     def __feature_encoder(self, image):
-#         """ features for correlation volume """
-#         return self.fnet(image).squeeze(0)
-
-#     @torch.cuda.amp.autocast(enabled=True)
-#     @torch.no_grad()
-#     def track(self, tstamp, image, depth=None, intrinsics=None):
-#         """ main update operation - run on every frame in video """
-
-#         Id = lietorch.SE3.Identity(1,).data.squeeze()
-#         ht = image.shape[-2] // 8
-#         wd = image.shape[-1] // 8
-
-#         # normalize images
-#         inputs = image[None, None, [2,1,0]].to(self.device) / 255.0
-#         inputs = inputs.sub_(self.MEAN).div_(self.STDV)
-
-#         # extract features
-#         gmap = self.__feature_encoder(inputs)
-
-#         ### always add first frame to the depth video ###
-#         if self.video.counter.value == 0:
-#             net, inp = self.__context_encoder(inputs)
-#             self.net, self.inp, self.fmap = net, inp, gmap
-#             self.video.append(tstamp, image, Id, 1.0, intrinsics / 8.0, gmap[0], net[0], inp[0])
-
-#         ### only add new frame if there is enough motion ###
-#         else:                
-#             # index correlation volume
-#             coords0 = pops.coords_grid(ht, wd, device=self.device)[None,None]
-#             corr = CorrBlock(self.fmap[None], gmap[None])(coords0)
-
-#             # approximate flow magnitude using 1 update iteration
-#             _, delta, weight = self.update(self.net[None], self.inp[None], corr)
-
-#             # check motion magnitue / add new frame to video
-#             if delta.norm(dim=-1).mean().item() > self.thresh:
-#                 self.count = 0
-#                 net, inp = self.__context_encoder(inputs)
-#                 self.net, self.inp, self.fmap = net, inp, gmap
-#                 self.video.append(tstamp, image, None, None, intrinsics / 8.0, gmap[0], net[0], inp[0])
-
-#             else:
-#                 self.count += 1
-
+                if custom: return False
\ No newline at end of file
diff --git a/setup.py b/setup.py
index 9af97d4..8f00220 100644
--- a/setup.py
+++ b/setup.py
@@ -24,6 +24,8 @@ setup(
                     '-gencode=arch=compute_75,code=sm_75',
                     '-gencode=arch=compute_80,code=sm_80',
                     '-gencode=arch=compute_86,code=sm_86',
+                    '-gencode=arch=compute_89,code=sm_89',
+                    '-gencode=arch=compute_90,code=sm_90',
                 ]
             }),
     ],
@@ -53,7 +55,9 @@ setup(
                     '-gencode=arch=compute_70,code=sm_70', 
                     '-gencode=arch=compute_75,code=sm_75',
                     '-gencode=arch=compute_80,code=sm_80',
-                    '-gencode=arch=compute_86,code=sm_86',                 
+                    '-gencode=arch=compute_86,code=sm_86',
+                    '-gencode=arch=compute_89,code=sm_89',
+                    '-gencode=arch=compute_90,code=sm_90',
                 ]
             }),
     ],
diff --git a/src/altcorr_kernel.cu b/src/altcorr_kernel.cu
index f418f5c..b53be29 100644
--- a/src/altcorr_kernel.cu
+++ b/src/altcorr_kernel.cu
@@ -306,7 +306,7 @@ std::vector<torch::Tensor> altcorr_cuda_forward(
   const dim3 threads(BLOCK_H, BLOCK_W);
 
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(fmap1.type(), "altcorr_forward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(fmap1.scalar_type(), "altcorr_forward_kernel", ([&] {
     altcorr_forward_kernel<scalar_t><<<blocks, threads>>>(
         fmap1.packed_accessor32<scalar_t,4,torch::RestrictPtrTraits>(),
         fmap2.packed_accessor32<scalar_t,4,torch::RestrictPtrTraits>(),
diff --git a/src/correlation_kernels.cu b/src/correlation_kernels.cu
index 2812067..df5735c 100644
--- a/src/correlation_kernels.cu
+++ b/src/correlation_kernels.cu
@@ -142,7 +142,7 @@ std::vector<torch::Tensor> corr_index_cuda_forward(
   torch::Tensor corr = torch::zeros(
     {batch_size, 2*radius+1, 2*radius+1, ht, wd}, opts);
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_forward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_forward_kernel", ([&] {
     corr_index_forward_kernel<scalar_t><<<blocks, threads>>>(
       volume.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
@@ -173,7 +173,7 @@ std::vector<torch::Tensor> corr_index_cuda_backward(
   const dim3 threads(BLOCK, BLOCK);
 
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.type(), "sampler_backward_kernel", ([&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(volume.scalar_type(), "sampler_backward_kernel", ([&] {
     corr_index_backward_kernel<scalar_t><<<blocks, threads>>>(
       coords.packed_accessor32<float,4,torch::RestrictPtrTraits>(),
       corr_grad.packed_accessor32<scalar_t,5,torch::RestrictPtrTraits>(),
diff --git a/src/droid_kernels.cu b/src/droid_kernels.cu
index a770112..5068690 100644
--- a/src/droid_kernels.cu
+++ b/src/droid_kernels.cu
@@ -2000,7 +2000,7 @@ class SparseBlockAsym {
           .dtype(torch::kFloat64)).to(torch::kCUDA).to(torch::kFloat32);
       }
       else {
-        std::cout << "Warning: Cholesky decomposition was not successful." << std::endl;
+        //std::cout << "Warning: Cholesky decomposition was not successful." << std::endl;
         dx = torch::zeros({N}, torch::TensorOptions()
           .device(torch::kCUDA).dtype(torch::kFloat32));
       }
